statistics/cmsketch.go

```diff
@@ -44,29 +60,176 @@ func NewCMSketch(d, w int32) *CMSketch {
 	return &CMSketch{depth: d, width: w, table: tbl}
 }
 
+// topNHelper wraps some variables used when building cmsketch with top n.
+type topNHelper struct {
+	sampleSize    uint64
+	numTop        uint32
+	counter       map[hack.MutableString]uint64
+	sorted        []uint64
+	onlyOnceItems uint64
+	sumTopN       uint64
+	lastVal       uint64
+}
+
+func newTopNHelper(sample [][]byte, numTop uint32) *topNHelper {
+	counter := make(map[hack.MutableString]uint64)
+	for i := range sample {
+		counter[hack.String(sample[i])]++
+	}
+	sorted, onlyOnceItems := make([]uint64, 0, len(counter)), uint64(0)
+	for _, cnt := range counter {
+		sorted = append(sorted, cnt)
+		if cnt == 1 {
+			onlyOnceItems++
+		}
+	}
+	sort.Slice(sorted, func(i, j int) bool {
+		return sorted[i] > sorted[j]
+	})
+
+	var (
+		// last is the last element in top N index should occurres atleast `last` times.
+		last      uint64
+		sumTopN   uint64
+		sampleNDV = uint32(len(sorted))
+	)
+	numTop = mathutil.MinUint32(sampleNDV, numTop) // In case numTop is bigger than sample NUV.
+	// The following loop builds find how many elements be added to the top N index.
+	// The final Top-N index may have at most 2*numTop elements for some less skewed sample.
+	for i := uint32(0); i < sampleNDV && i < numTop*2; i++ {
+		// Here, 2/3 is get by running tests, tested 1, 1/2, 2/3, and 2/3 is relative better than 1 and 1/2.
+		// If the frequency of i-th elements is close to n-th element, it is added to the top N index too.
```

Comment:
how about moving this comment on the top of the `for` loop and replace it with:
```go
    // Only element whose frequency is not smaller than 2/3 multiples the
    // frequency of the n-th element are added to the TopN statistics. We chose
    // 2/3 as an empirical value because the average cardinality estimation
    // error is relatively small compared with 1, 1/2 and 1/3.
```